### PR TITLE
Warn (don't fail!) on spelling errors. Fix typos reported by codespell.

### PR DESCRIPTION
**Pull request by practicalswift**
* Check for common misspellings using `codespell`.
* Fix recently introduced typos reported by `codespell`.

### CODE REVIEW

**Comment by fanquake** [comment]
Travis failure:
```
33.63s$ test/lint/lint-all.sh
test/lint/lint-all.sh: line 21: test/lint/lint-spelling.sh: Permission denied
^---- failure generated from test/lint/lint-spelling.sh
```

File: .travis.yml [hunk 1]
```diff
@@ -143,11 +143,11 @@ jobs:
         BITCOIN_CONFIG="--enable-gui --enable-reduce-exports --enable-werror"
     - stage: lint
       env:
-      sudo: false
       cache: false
       language: python
       python: '3.6'
       install:
+        - travis_retry sudo apt-get install --no-install-recommends --no-upgrade -qq codespell
```
Comment (ken2812221) [inline .travis.yml]: Can use
```
addons:
  apt:
    packages:
    - codespell
```
to avoid sudo

Comment (practicalswift) [reply .travis.yml]: Oh that is much cleaner! Thanks!

**Comment by laanwj** [comment]
Does this mean we need to manually add words that are not recognized to the spell checker?
If we do this, I think this certainly needs maintenance instructions.

File: test/lint/lint-spelling.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+#
+# Check for spelling errors.
+
+export LC_ALL=C
+
+EXIT_CODE=0
+EXCLUDE_REGEXP="(src/example.cpp:.*example-false-positive-word) "
+SPELLING_ERRORS=$(codespell -d -q 3 $(git ls-files -- ":(exclude)src/leveldb/" ":(exclude)src/secp256k1/" ":(exclude)src/univalue/" ":(exclude)depends/" ":(exclude)doc/release-notes/" ":(exclude)src/qt/locale/" ":(exclude)test/functional/test_framework/__init__.py" ":(exclude)test/lint/lint-spelling.sh") | grep -vE "${EXCLUDE_REGEXP}")
```
Comment (Empact) [inline test/lint/lint-spelling.sh]: nit: using the long form options here would be more self-documenting: `--disable-colors --quiet-level=`

File: src/rpc/net.cpp [hunk 3]
```diff
@@ -242,13 +242,13 @@ static UniValue disconnectnode(const JSONRPCRequest& request)
...
         throw std::runtime_error(
             "disconnectnode \"[address]\" [nodeid]\n"
-            "\nImmediately disconnects from the specified peer node.\n"
-            "\nStrictly one out of 'address' and 'nodeid' can be provided to identify the node.\n"
-            "\nTo disconnect by nodeid, either set 'address' to the empty string, or call using the named 'nodeid' argument only.\n"
-            "\nArguments:\n"
+            "\n" "Immediately disconnects from the specified peer node.\n"
+            "\n" "Strictly one out of 'address' and 'nodeid' can be provided to identify the node.\n"
+            "\n" "To disconnect by nodeid, either set 'address' to the empty string, or call using the named 'nodeid' argument only.\n"
+            "\n" "Arguments:\n"
```
Comment (Empact) [inline src/rpc/net.cpp]: Why are these changes necessary? There are plenty more leading `\n` in this file.

File: .travis.yml [hunk 4]
```diff
@@ -147,6 +147,10 @@ jobs:
       cache: false
       language: python
       python: '3.6'
+      addons:
+        apt:
+          packages:
+          - codespell
```
Comment (Empact) [inline .travis.yml]: `codespell` recommends installation via `pip`:
https://github.com/codespell-project/codespell

**Comment by Empact** [comment]
Concept ACK

File: test/lint/lint-spelling.sh (hunk 2 above)
Comment (practicalswift) [reply test/lint/lint-spelling.sh]: Good point! Now fixed.

File: src/rpc/net.cpp (hunk 3 above)
Comment (practicalswift) [reply src/rpc/net.cpp]: The old Ubuntu packaged version didn't understand `\n` and thus flagged "nTo" as a typo. That seems to have been fixed in the more recent `pip` installed version.

File: .travis.yml (hunk 4 above)
Comment (practicalswift) [reply .travis.yml]: Good point! Now fixed

File: src/qt/paymentserver.h [hunk 5]
```diff
@@ -77,7 +77,7 @@ class PaymentServer : public QObject
     // to read from the file specified in the -rootcertificates setting,
     // or, if that's not set, to use the system default root certificates.
     // If you pass in a store, you should not X509_STORE_free it: it will be
-    // freed either at exit or when another set of CAs are loaded.
+    // freed either at exit or when another set of CA:s are loaded.
```
Comment (MarcoFalke) [inline src/qt/paymentserver.h]: why?

File: src/qt/coincontroldialog.cpp [hunk 6]
```diff
@@ -185,7 +185,7 @@ void CoinControlDialog::buttonSelectAllClicked()
                 ui->treeWidget->topLevelItem(i)->setCheckState(COLUMN_CHECKBOX, state);
     ui->treeWidget->setEnabled(true);
     if (state == Qt::Unchecked)
-        coinControl()->UnSelectAll(); // just to be sure
+        coinControl()->DeSelectAll(); // just to be sure
```
Comment (MarcoFalke) [inline src/qt/coincontroldialog.cpp]: Should probably be a scripted diff

File: src/qt/paymentserver.h (hunk 5 above)
Comment (Empact) [reply src/qt/paymentserver.h]: Yeah I would add these to `test/lint/lint-spelling.ignore-words.txt` rather than use `:` to separate them.

**Comment by laanwj** [comment]
I'm... not convinced about this.

I think running this once in a while is good, just to see if there's any really awkward misspellings, but running this in travis and failing the tests just because of a misspelled word in a comment seems overkill.

As I've said before, many times, linters should be added cautiously, there to prevent real problems resulting in bugs.

**Comment by practicalswift** [comment]
@laanwj The reason I wrote this linter is your comment in #9962 which made clear that your preference was to catch typos as review time so that they never get merged in the first place, and that this should be done without involving the maintainers.

From https://github.com/bitcoin/bitcoin/pull/9962#issuecomment-285332911:

> In principle it is good to fix comment typos, but my preference too would be to either catch them at the source (review the PR and find them) so that they never get merged in the first place or bunch them up into something like a "fix typos" pull every few weeks.

From https://github.com/bitcoin/bitcoin/pull/9962#issuecomment-285334709:

> […] This allows live cooperation on fixing typos without involving the maintainers at every step.

Also @fanquake requested an automated solution in #10419:

> Im curious. What tool are you using to find these? I assume it's being run in some automated way on the repository when new changes are merged? Why not merge the open pull requests instead, catch the typos/incorrect includes etc etc and then advise changes in the PRs while they are still open?

**Comment by scravy** [comment]
Maybe it would make sense to have this run as a separate check (as in github check – https://developer.github.com/v3/checks/runs/ ). Especially with things like an AppVeyor build for MSVC (https://github.com/bitcoin/bitcoin/pull/13964#issuecomment-412852328, which already is a different check) it might be handy to have this split out of the other builds.

It would need to be hosted somewhere though I guess.

**Comment by practicalswift** [comment]
@laanwj Can you clarify your position? :-)

Instead of exiting with a failure if new typos are introduced in a submitted PR we could perhaps just let `codespell` print the identified typos to the console as part of the linting process. Sounds like a good compromise?

**Comment by practicalswift** [comment]
Now only warns in case of spelling errors in order to not Travis. Also rebased :-)

**Comment by practicalswift** [comment]
The output can be seen on https://travis-ci.org/bitcoin/bitcoin/jobs/421142979:

```
…
$ set -o errexit; source .travis/lint_06_script.sh
Running script for: 52bc0698ca0b27cd24f6f31b335495d641fd58cf
sed -i 's/UnSelect/DeSelect/g' src/qt/coincontroldialog.cpp src/qt/sendcoinsdialog.cpp src/wallet/coincontrol.h
OK
src/crypto/ctaes in HEAD currently refers to tree 1b6c31139a71f80245c09597c343936a8e41d021
src/crypto/ctaes in HEAD was last updated in commit 8501bedd7508ac514385806e191aec21ee978891 (tree 1b6c31139a71f80245c09597c343936a8e41d021)
subtree commit 003a4acfc273932ab8c2e276cde3b4f3541012dd unavailable: cannot compare
src/secp256k1 in HEAD currently refers to tree af619602c243e0d8fbd5934f375faa4aedb4ca6e
src/secp256k1 in HEAD was last updated in commit fd86f998fcfd25d823d67a2920814e22445655f9 (tree af619602c243e0d8fbd5934f375faa4aedb4ca6e)
subtree commit 0b7024185045a49a1a6a4c5615bf31c94f63d9c4 unavailable: cannot compare
src/univalue in HEAD currently refers to tree a2dbaf30021c0a31eecc213cd84d600273d333b6
src/univalue in HEAD was last updated in commit a570098021be6a7b9f4589300ea655ae4633628e (tree a2dbaf30021c0a31eecc213cd84d600273d333b6)
subtree commit 51d3ab34ba2857f0d03dc07250cb4a2b5e712e67 unavailable: cannot compare
src/leveldb in HEAD currently refers to tree f8cdc77167be86194f98d0412baf4b89d3d5abe6
src/leveldb in HEAD was last updated in commit ec749b1bcdf2483b642fb51d635800e272c68ba6 (tree f8cdc77167be86194f98d0412baf4b89d3d5abe6)
subtree commit 524b7e36a8e3bce6fcbcd1b5df09024283f325ba unavailable: cannot compare
Args used        : 181
Args documented  : 186
Args undocumented: 0
set()
Args unknown     : 5
{'-zmqpubhashtx', '-zmqpubhashblock', '-zmqpubrawtx', '-promiscuousmempoolflags', '-zmqpubrawblock'}
* Checking consistency between dispatch tables and vRPCConvertParams
src/streams.h:532: signficant  ==> significant
^ Warning: codespell identified likely spelling errors. Any false positives? Add them to the list of ignored words in test/lint/lint-spelling.ignore-words.txt
The command "set -o errexit; source .travis/lint_06_script.sh" exited with 0.
…
```

Pretty sweet :-)

**Comment by practicalswift** [comment]
Rebased and fixed the typos that `codespell` found in the newly introduced documentation.

`codespell` catched the "account" typo first reported by @promag in PR #14023 post-merge: https://github.com/bitcoin/bitcoin/pull/14023#discussion_r213136227

File: src/dbwrapper.cpp [hunk 7]
```diff
@@ -78,7 +78,7 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
     // do not interfere with select() loops. On 64-bit Unix hosts this value is
     // also OK, because up to that amount LevelDB will use an mmap
     // implementation that does not use extra file descriptors (the fds are
-    // closed after being mmaped).
+    // closed after being mmap:ed).
```
Comment (jnewbery) [inline src/dbwrapper.cpp]: I've never seen a colon used in this way. Why not mmap'ed?

File: src/qt/coincontroldialog.cpp (hunk 6 above)
Comment (jnewbery) [inline src/qt/coincontroldialog.cpp]: Is unselect really not ok? https://en.oxforddictionaries.com/definition/unselect.

**Review by jnewbery — COMMENTED** [review]
Apart from the unselect->deselect change, I think the spelling corrections are fine.

Concept ACK on the linting, although I haven't reviewed the code. I'd lean toward having the linter fail on spelling errors rather than just warn (which will no doubt be ignored).

File: src/qt/coincontroldialog.cpp (hunk 6 above)
Comment (practicalswift) [reply src/qt/coincontroldialog.cpp]: @jnewbery I found this: https://english.stackexchange.com/questions/18465/unselect-or-deselect

I'm happy to revert this specific change if it is controversial :-)

**Comment by practicalswift** [comment]
@jnewbery Thanks for the encouraging review! PR updated – please re-review :-)

If this PR is merged I'll look if there is some smart way to use Travis folding or similar to make the spelling warnings more prominent in the Travis interface.

**Comment by laanwj** [comment]
>  I'd lean toward having the linter fail on spelling errors rather than just warn (which will no doubt be ignored).

I still strongly disagree with this, I don't think we should pile up linters for cosmetic issues.

The original idea of them was to flag serious issues that can result in bugs.

Failing the tests on some supposed spelling error in a comment really, goes too far.

**Comment by practicalswift** [comment]
@laanwj What about enabling it on a warning level (as the PR is currently formulated)? Could that be a good compromise? :-)

**Comment by jnewbery** [comment]
> I still strongly disagree with this

ok, I won't rehash the debate that we've all had in countless other PRs already :)

My point was merely that warnings tend to get ignored, so in general I think they're less useful than errors. I don't have any strong opinion on this PR though.

**Comment by practicalswift** [comment]
@jnewbery When thinking about it again I agree regarding "deselect" vs "unselect". Now added it to the ignore list instead and removed that commit. Please re-review :-)

**Comment by Sjors** [comment]
Concept ACK for warnings. Maybe Drahbot can point out typos. In addition we could fix the ones we missed as part of the release process.

utACK 4a98785 (though maybe squash the first and last commit)

**Comment by practicalswift** [comment]
@Sjors Squashed. Please re-review :-)

**Comment by Sjors** [comment]
utACK b5ee29e460dc36066bac68e041073100c0eca20e

File: src/qt/coincontroldialog.cpp [hunk 8]
```diff
@@ -441,7 +441,7 @@ void CoinControlDialog::updateLabels(WalletModel *model, QDialog* dialog)
     for (const auto& out : model->wallet().getCoins(vCoinControl)) {
         if (out.depth_in_main_chain < 0) continue;
 
-        // unselect already spent, very unlikely scenario, this could happen
+        // deselect already spent, very unlikely scenario, this could happen
```
Comment (fanquake) [inline src/qt/coincontroldialog.cpp]: If unselect is in the ignore list, why is this being changed?

Comment (practicalswift) [reply src/qt/coincontroldialog.cpp]: Good point. Fixed!

**Comment by practicalswift** [comment]
@Sjors @fanquake Updated. Please re-review :-)

**Comment by MarcoFalke** [comment]
utACK f8a81f73ac4fee82488806c2b384b36e8ee2acdf

File: test/lint/lint-spelling.ignore-words.txt [hunk 9]
```diff
@@ -0,0 +1,6 @@
+cas
+hights
```
Comment (Empact) [inline test/lint/lint-spelling.ignore-words.txt]: I don't find `hights` in current master

Comment (practicalswift) [reply test/lint/lint-spelling.ignore-words.txt]: @Empact 

```
$ git grep -i hights
contrib/linearize/linearize-data.py:        self.highTS = 1408893517 - 315360000
contrib/linearize/linearize-data.py:                os.utime(self.outFname, (int(time.time()), self.highTS))
contrib/linearize/linearize-data.py:                    os.utime(self.outFname, (int(time.time()), self.highTS))
contrib/linearize/linearize-data.py:        if blkTS > self.highTS:
contrib/linearize/linearize-data.py:            self.highTS = blkTS
```

:-)

**Comment by Empact** [comment]
utACK ada3562

**Comment by practicalswift** [comment]
@laanwj Oh, I wasn't aware of the policy change introduced in #14179 so I don't have anything meaningful to add about that. Setting the discussion about that merge decision aside:

Do you have a link to the Travis failure? It would be interesting to see if it was a false positive or a false positive that caused the Travis failure.

**Comment by luke-jr** [comment]
It came up when someone tried to name a mutex `mut`. Seems unreasonable to demand variables have English word names...

**Comment by laanwj** [comment]
yes it came up on IRC @karel-3d
```
08:29 < karelb> I am trying to run my fork of bitcoind on travis.... and travis stops with this
08:29 < karelb> src/threadinterrupt.cpp:25: mut  ==> must, mutt, moot
08:29 < karelb> Warning: codespell identified likely spelling errors
08:29 < karelb> failure generated from test/lint/lint-spelling.sh
08:29 < karelb> .....ummmmm, ok?
08:30 < karelb> I did no change in threadinterrupt, and that "mut" is just some mutex
08:39 < karelb> well whatever, one commit that renamed "mut" to "mutex" fixed that, I just wonder why I needed to do that...
08:40 < gwillen> karelb: your travis instance is stupid
08:40 < gwillen> if you make it less stupid you will have fewer problems
08:40 < gwillen> it is trying to spell-correct your code, and doing it moronically
08:40 < karelb> :D
08:40 < gwillen> and seems to be set to warnings-as-errors.
08:40 < karelb> I use travis-ci.org
08:41 < gwillen> well, apparently they are stupid
08:41 < gwillen> but I imagine there is an option not to make codespell warnings errors
08:41 < karelb> the same config as bitcoin core
08:41 < gwillen> which they absolutely should not be
08:41 < gwillen> oh, hm
08:41 < gwillen> I mean, seemingly not
08:42 < karelb> well let's see what happens when I make a PR, if that `mut` stuff is still there
10:49 < wumpus> karelb: lol that's stupid
10:50 < wumpus> please don't tell me that stupid spelling check is mandatory in travis now
```

**Comment by scravy** [comment]
@practicalswift Which versions of codespell?

**Comment by practicalswift** [comment]
@scravy 

```
$ pip3 install codespell --upgrade
Requirement already up-to-date: codespell in /usr/local/lib/python3.6/dist-packages
$ codespell --version
1.14.0
```

**Comment by practicalswift** [comment]
Fixed in #14495 :-)

**Comment by practicalswift** [comment]
@scravy Seems like this was introduced somewhere between `1.13.0` and `1.14.0`. We're pinning to `1.13.0` in Travis so I don't know how `1.13.0` got installed in @karel-3d:s Travis setup. @karel-3d, do you know? :-)

**Comment by karel-3d** [comment]
I cloned the repo, did some experimental changes, and used Travis-ci.org.

The build is here

https://travis-ci.org/karel-3d/bitcoin/builds/442019337

It's branched from this commit, 18 days old. (Maybe it will go away if I rebase on current master?)

https://github.com/bitcoin/bitcoin/commit/0809e68a9084630ff4d11ec0503a3d0ab52bc6d7

**Comment by karel-3d** [comment]
Oh yeah. It seems I have branched the master *before* this commit

https://github.com/bitcoin/bitcoin/commit/d10f2cd7d8506aec8b406c346b35d3c31f0f37df

Which means that if I rebase to master, it would indeed go away.

**Comment by practicalswift** [comment]
@karel-3d Thanks for the clarification! That explains the mystery :-)
